### PR TITLE
Range-ify std.file.read()

### PR DESCRIPTION
**Pull request by WalterBright**
Requires https://github.com/D-Programming-Language/phobos/pull/3417

http://dlang.org/phobos/std_file.html#.read

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -163,6 +163,44 @@ private T cenforce(T)(T condition, lazy const(char)[] name, string file = __FILE
     return condition;
 }
 
+version (Windows)
+@trusted
+private T cenforce(T)(T condition, const(char)[] name, const(wchar)* namez, string file = __FILE__, size_t line = __LINE__)
+{
+    if (!condition)
```
Comment (andralex) [inline std/file.d]: Short path first:

```d
if (condition)
    return condition;
...
```

File: std/file.d [hunk 2]
```diff
@@ -188,7 +226,15 @@ Returns: Untyped array of bytes _read.
...
-version (Posix) void[] read(in char[] name, size_t upTo = size_t.max) @trusted
+version (Posix) void[] read(R)(R name, size_t upTo = size_t.max)
+{
+    static if (isNarrowString!R && is(Unqual!(ElementEncodingType!R) == char))
+        return readImpl(name, name.tempCString(), upTo);
+    else
+        return readImpl(null, name.tempCString(), upTo);
+}
+
+version (Posix) private void[] readImpl(const(char)[] name, const(char)* namez, size_t upTo = size_t.max) @trusted
```
Comment (andralex) [inline std/file.d]: this looks like it best goes `static nested` inside of `read`

File: std/file.d [hunk 3]
```diff
@@ -234,15 +280,25 @@ version (Posix) void[] read(in char[] name, size_t upTo = size_t.max) @trusted
...
+    if ((isInputRange!R && isSomeChar!(ElementEncodingType!R)) || isSomeString!R)
+{
+    static if (isNarrowString!R && is(Unqual!(ElementEncodingType!R) == char))
+        return readImpl(name, name.tempCString!wchar(), upTo);
+    else
+        return readImpl(null, name.tempCString!wchar(), upTo);
+}
+
+
+version (Windows) private void[] readImpl(const(char)[] name, const(wchar)* namez, size_t upTo = size_t.max) @safe
```
Comment (andralex) [inline std/file.d]: same here - nest this inside `read`

File: std/file.d (hunk 2 above)
Comment (WalterBright) [inline std/file.d]: I wanted it outside as a non-template function so it will not get repeatedly generated.

File: std/file.d (hunk 3 above)
Comment (WalterBright) [inline std/file.d]: same answer!

File: std/file.d (hunk 2 above)
Comment (andralex) [inline std/file.d]: urgh

File: std/file.d (hunk 3 above)
Comment (andralex) [inline std/file.d]: same urgh :)

**Comment by andralex** [comment]
mkay

**Pull request merged**